### PR TITLE
[10.x.x] Include PostProcessing toggle added to the URP Asset

### PR DESCRIPTION
**Pull request by ellioman**
# Purpose of this PR
This PR does the following:
- Adds a toggle in the URP Asset to include or not Post-Processing
- Adds a warning to the Camera Inspector when PostProcessing is enabled on a camera but the include toggle is turned off
- Strips Post-Processing shaders from builds if the toggle is off
- Moves the PostProcessData variable from renderers to the Asset
- Adds a small helper script to the URP Graphics Test Project to print asset GUIDS
- Moves the CopyImageToReferenceFolders script to /CommonAssets/Editor

image
image
image
image

---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

# Yamato
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Ffeature%252Fdisable-post-processing

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs [hunk 1]
```diff
@@ -174,7 +174,9 @@ public static void RenderSingleCamera(ScriptableRenderContext context, Camera ca
             if (asset.useAdaptivePerformance)
                 ApplyAdaptivePerformance(ref cameraData);
 #endif
-            RenderSingleCamera(context, cameraData, cameraData.postProcessEnabled);
+
+            bool anyPostProcessing = cameraData.postProcessEnabled && asset.postProcessIncluded;
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs:178]: I can totally see RenderingSingleCamera(ctx, cam, bool, bool, bool, bool, bool, bool, bool, ... ) being thing in the future. Should we have a settings/options struct, pass in the asset and ask from it or at least do an enum i.e. that would read like PostProcess.Enabled/.Disabled?

**Comment by CaptainKirby** [comment]
Did a quick run through of the PR, check the card [here ](https://favro.com/organization/c564ede4ed3337f7b17986b6/1b0fabd725486d71d77e9c1f?card=Uni-142748)if you want, but everything seems to work as expected :)

**Comment by oleks-k** [comment]
On the camera, the checkbox is called "Post-Processing", and on the URP asset it's "Post-processing". I think it should be the same.
On the Camera, the check-box, in "Anti-aliasing" the second word is not capitalized. So should it be "Post-processing" everywhere?
I'm not sure since I'm not aware of the rules that the UX team is using. @ellioman , do you know if we have some formal rules for naming UX elements?

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineCameraEditor.cs [hunk 2]
```diff
@@ -59,6 +59,7 @@ static class Styles
 
             public static readonly string missingRendererWarning = "The currently selected Renderer is missing form the Universal Render Pipeline asset.";
             public static readonly string noRendererError = "There are no valid Renderers available on the Universal Render Pipeline asset.";
+            public static readonly string disabledPostprocessing = "Post-processing is currently disabled on the current Universal Render Pipeline asset.";
```
Comment (martint-unity) [inline com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineCameraEditor.cs:62]: capital P to keep with consistency.

**Comment by martint-unity** [comment]
Also when unclicking Post Processing on a Camera the Warning is still there. This should also be fixed before merging.
image

**Comment by martint-unity** [comment]
Talked to UX/UI and please make it Post Processing.  no - in between.

File: TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
+
+public class PrintAssetGUID
+{
+    // Add a new menu item that is accessed by right-clicking on an asset in the project view
+
+    [MenuItem("Assets/PrintAssetGUID", priority = 2)]
+    private static void CopyImages()
```
[the remark this answers is not in the document]
Comment (ellioman) [reply TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs:13]: Deleted the file.

File: TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        StringBuilder sb = new StringBuilder(Selection.objects.Length * 20);
+        for (int i = 0; i < Selection.objects.Length; i++)
+        {
+            var path = AssetDatabase.GetAssetPath(Selection.objects[i]);
+            var guid = AssetDatabase.AssetPathToGUID(path);
+            sb.AppendLine(guid + " : " + path);
+        }
+        Debug.Log(sb);
+    }
+}
```
[the remark this answers is not in the document]
Comment (ellioman) [reply TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs:24]: Deleted the file.

File: TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+using UnityEngine;
+using UnityEditor;
+using System.IO;
+using System.Collections.Generic;
+using System.Text;
+using System.Linq;
```
[the remark this answers is not in the document]
Comment (ellioman) [reply TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs:6]: Deleted the file.

File: TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
+using UnityEngine;
+using UnityEditor;
+using System.IO;
```
[the remark this answers is not in the document]
Comment (ellioman) [reply TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs:3]: Deleted the file

File: TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+using System.Linq;
+
+public class PrintAssetGUID
+{
+    // Add a new menu item that is accessed by right-clicking on an asset in the project view
+
+    [MenuItem("Assets/PrintAssetGUID", priority = 2)]
+    private static void CopyImages()
+    {
+        StringBuilder sb = new StringBuilder(Selection.objects.Length * 20);
```
[the remark this answers is not in the document]
Comment (ellioman) [reply TestProjects/UniversalGraphicsTest/Assets/CommonAssets/Editor/PrintAssetGUID.cs:15]: Deleted the file. We're gonna implement in to the native side :)

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs [hunk 8]
```diff
@@ -56,6 +56,7 @@ internal class Styles
             public static GUIContent supportsSoftShadows = EditorGUIUtility.TrTextContent("Soft Shadows", "If enabled pipeline will perform shadow filtering. Otherwise all lights that cast shadows will fallback to perform a single shadow sample.");
 
             // Post-processing
+            public static GUIContent postProcessInclude = EditorGUIUtility.TrTextContent("Include Post-processing", "Includes or removes Post-processing from builds, this includes rendering passes, shaders and texture resources.");
```
[the remark this answers is not in the document]
Comment (ellioman) [reply com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAssetEditor.cs:59]: I take it you mean the "Post-processing" in this comment?
I've fixed that now to "Post Processing"

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineCameraEditor.cs (hunk 2 above)
Comment (ellioman) [reply com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineCameraEditor.cs:62]: Done!

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs (hunk 1 above)
Comment (ellioman) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs:183]: Makes perfect sense. I would like to do that in a separate PR though.

File: com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs [hunk 9]
```diff
@@ -43,15 +43,44 @@ internal class ShaderPreprocessor : IPreprocessShaders
...
-            if (!CoreUtils.HasFlag(features, ShaderFeatures.MainLightShadows) &&
-                shader.name.Contains("ScreenSpaceShadows"))
+            if (!CoreUtils.HasFlag(features, ShaderFeatures.MainLightShadows) && shader.name.Contains("ScreenSpaceShadows"))
+            {
                 return true;
+            }
+
+            if (!urpAsset.postProcessIncluded)
+            {
+                var path = AssetDatabase.GetAssetPath(shader);
```
[the remark this answers is not in the document]
Comment (ellioman) [reply com.unity.render-pipelines.universal/Editor/ShaderPreprocessor.cs:79]: Yep! Fixed.

**Comment by ellioman** [comment]
Closing this PR for now. Will be reopened and redone at a later point.

**Pull request closed without merging**